src/Servers/Kestrel/Core/src/Internal/Infrastructure/KestrelMetrics.cs

```diff
@@ -287,19 +300,92 @@ private void TlsHandshakeStopCore(in ConnectionMetricsContext metricsContext, lo
+            tags.Add("server.port", localIPEndPoint.Port);
+
+            switch (localIPEndPoint.Address.AddressFamily)
+            {
+                case AddressFamily.InterNetwork:
+                    tags.Add("network.type", "ipv4");
+                    break;
+                case AddressFamily.InterNetworkV6:
+                    tags.Add("network.type", "ipv6");
+                    break;
+            }
+
+            // There isn't an easy way to detect whether QUIC is the underlying transport.
+            // This code assumes that a multiplexed connection is QUIC.
+            // Improve in the future if there are additional multiplexed connection types.
+            var transport = metricsContext.ConnectionContext is not MultiplexedConnectionContext ? "tcp" : "udp";
+            tags.Add("network.transport", transport);
+        }
+        else if (localEndpoint is UnixDomainSocketEndPoint udsEndPoint)
         {
-            // TODO: Improve getting string allocation for endpoint. Currently allocates.
-            // Considering adding a way to cache on ConnectionContext.
-            tags.Add("endpoint", localEndpoint.ToString());
+            tags.Add("server.address", udsEndPoint.ToString());
+            tags.Add("network.transport", "unix");
+        }
+        else if (localEndpoint is NamedPipeEndPoint namedPipeEndPoint)
+        {
+            tags.Add("server.address", namedPipeEndPoint.ToString());
+            tags.Add("network.transport", "pipe");
+        }
+        else if (localEndpoint != null)
+        {
+            tags.Add("server.address", localEndpoint.ToString());
+            tags.Add("network.transport", localEndpoint.AddressFamily.ToString());
         }
     }
 
     public ConnectionMetricsContext CreateContext(BaseConnectionContext connection)
     {
         // Cache the state at the start of the connection so we produce consistent start/stop events.
         return new ConnectionMetricsContext(connection,
-            _currentConnectionsCounter.Enabled, _connectionDuration.Enabled, _queuedConnectionsCounter.Enabled,
-            _queuedRequestsCounter.Enabled, _currentUpgradedRequestsCounter.Enabled, _currentTlsHandshakesCounter.Enabled);
+            _activeConnectionsCounter.Enabled, _connectionDuration.Enabled, _queuedConnectionsCounter.Enabled,
+            _queuedRequestsCounter.Enabled, _currentUpgradedRequestsCounter.Enabled, _activeTlsHandshakesCounter.Enabled);
+    }
+
+    public static bool TryGetHandshakeProtocol(SslProtocols? protocols, [NotNullWhen(true)] out string? name, [NotNullWhen(true)] out string? version)
+    {
+        if (protocols != null)
+        {
+#pragma warning disable CS0618 // Type or member is obsolete
+#pragma warning disable SYSLIB0039 // Type or member is obsolete
+            switch (protocols.Value)
+            {
+                case SslProtocols.Ssl2:
+                    name = "ssl";
```

Comment:
do we still need SSL?

https://github.com/dotnet/aspnetcore/issues/49857:

> Do we need tls.protocol.name if using any SSL version with SslStream on dotnet 8 and above is impossible?

Reply:
I'm leaving it here in an abundance of caution. I'll leave a comment.